src/pos_kernel.cpp

```diff
@@ -14,15 +17,70 @@ namespace pos {
     }
 
     bool
-    CheckKernelHash(const uint256& stakeModifier, uint32_t nBits, int64_t height, int64_t coinstakeTime, const uint256& masternodeID, const Consensus::Params& params) {
+    CheckKernelHash(const uint256& stakeModifier, uint32_t nBits, int64_t height, int64_t coinstakeTime, uint64_t blockHeight, const uint256& masternodeID, const Consensus::Params& params) {
         // Base target
         arith_uint256 targetProofOfStake;
         targetProofOfStake.SetCompact(nBits);
 
         const auto hashProofOfStake = UintToArith256(CalcKernelHash(stakeModifier, height, coinstakeTime, masternodeID, params));
 
+        // New difficulty calculation to make staking easier the longer it has
+        // been since a masternode staked a block.
+        if (blockHeight >= static_cast<uint64_t>(Params().GetConsensus().DakotaCrescentHeight))
+        {
+            auto node = pcustomcsview->GetMasternode(masternodeID);
+            if (!node)
+            {
+                return false;
+            }
+
+            // Default to min age
+            int64_t nTimeTx{params.pos.nStakeMinAge};
+
+            if (node->mintedBlocks > 0)
+            {
+                LOCK(cs_main);
+
+                const CBlockIndex* stakerBlock{nullptr};
+
+                // Loop until we find last staked block or hit max stake age.
+                for (const CBlockIndex* tip = ::ChainActive().Tip() ; tip && coinstakeTime - tip->GetBlockTime() < params.pos.nStakeMaxAge; tip = tip->pprev)
+                {
+                    CKeyID staker;
+                    if (tip->GetBlockHeader().ExtractMinterKey(staker))
+                    {
+                        auto id = pcustomcsview->GetMasternodeIdByOperator(staker);
+                        if (id && id == masternodeID)
+                        {
+                            // Found the last staked block
+                            stakerBlock = tip;
+                            break;
+                        }
+                    }
+                }
```

Comment:
How is the performance of this section of code? If consider the lookup for nStakeMaxAge which is 2 weeks, then roughly need to search for 32,000 blocks and each time need to call `ExtractMinterKey` and `GetMasternodeIdByOperator`, it may take some time. Maybe can consider save the staker information in cache, using lookup will be much faster.

Reply:
The performance is very bad, something like 2.7 seconds. Working on a DB look up instead.

Reply:
Also will save staker block in an in memory cache after loading from DB to save repeat lookups.

Reply:
We now store the masternode and time for every single block so the above code is now obsolete.